[mlir][Pass] Enable the option for reproducer generation without crashing

plotfi opened the pull request:
This PR adds API `makeReproducer` and cl::opt flag `--mlir-generate-reproducer=<filename>` in order to allow for mlir reproducer dumps even when the pipeline doesn't crash.

This PR also decouples the code that handles generation of an MLIR reproducer from the crash recovery portion. The purpose is to allow for generating reproducers outside of the context of a compiler crash.

This will be useful for frameworks and runtimes that use MLIR where it is needed to reproduce the pipeline behavior for reasons outside of diagnosing crashes. An example is for diagnosing performance issues using offline tools, where being able to dump the reproducer from a runtime compiler would be helpful.

joker-eph commented:
Being able to generate a reproducer for a framework (or other) outside of a crash seems useful, but it’s not clear to me why it is a passe manager option?
I would think that the framework can already generate it independently before calling the pass manager.

plotfi commented:
> Being able to generate a reproducer for a framework (or other) outside of a crash seems useful, but it’s not clear to me why it is a passe manager option?

I considered that, but I am primarily trying to target the mlir-opt's reproducer file ease of use here. Can I generate a reproducer file that could be loaded by mlir-opt or similar tools without doing a pass manager option?

plotfi commented:
@joker-eph Ah, do you mean then to invoke something like RecoveryReproducerContext::generate() from the runtime?

joker-eph commented:
> @joker-eph Ah, do you mean then to invoke something like RecoveryReproducerContext::generate() from the runtime?

Yes exactly!

If we need to extract the implementation of `RecoveryReproducerContext::generate()` as a more easily reusable API, that's an OK refactoring of course.

plotfi commented:
> > @joker-eph Ah, do you mean then to invoke something like RecoveryReproducerContext::generate() from the runtime?
> 
> Yes exactly!
> 
> If we need to extract the implementation of `RecoveryReproducerContext::generate()` as a more easily reusable API, that's an OK refactoring of course.

I will have to dive a little deeper into the current implementation to see how things could be setup without doing it in the pass manager directly. Ideally I'd want a similar mechanism to how the pass manager does it where you dump out the reproducer after everything has run. The pass manager's finalizer just made this kind of conveniently easy, but I am not certain if it is the right approach either.

plotfi commented:
> > @joker-eph Ah, do you mean then to invoke something like RecoveryReproducerContext::generate() from the runtime?
> 
> Yes exactly!
> 
> If we need to extract the implementation of `RecoveryReproducerContext::generate()` as a more easily reusable API, that's an OK refactoring of course.

Work in progress, but I have refactored the bits that append the reproducer info without having to do it from inside of the pass manager. The reproducers can be appended like so:

```
  for (auto &pass : pm.getPasses())
    makeReproducer(&pass, op.get(), "reproducer.mlir");
```

plotfi commented on mlir/test/Pass/crashless-reproducer.mlir:
```diff
@@ -0,0 +1,13 @@
+// RUN: mlir-opt %s -pass-pipeline='builtin.module(builtin.module(test-module-pass))' \
+// RUN:   -mlir-pass-pipeline-crash-reproducer=%t \
+// RUN:   -mlir-pass-pipeline-always-generate-reproducer=true -verify-diagnostics
```
This test is not working at the moment, working on it. I have something working manually on my local machine.

plotfi committed:
[NFC][mlir][Pass] Reorganize MLIR Crash Recovery Reproducer implementation

This patch decouples the code that handles generation of an MLIR reproducer
from the crash recovery portion. The purpose is to allow for generating
reproducers outside of the context of a compiler crash.

(cherry picked from commit bfb0682f2a318a84a05f91466757a03d68ddfd1f)

plotfi committed:
[mlir][Pass] Enable the option for reproducer generation without crashing

This patch adds API `makeReproducer` and cl::opt flag
--mlir-generate-reproducer=<filename> in order to allow for mlir
reproducer dumps even when the pipeline doesn't crash.

This will be useful for frameworks and runtimes that use MLIR where it is
needed to reproduce the pipeline behavior for reasons outside of diagnosing
crashes. An example is for diagnosing performance issues using offline tools,
where being able to dump the reproducer from a runtime compiler would be
helpful.

plotfi commented:
@joker-eph I've separated out the NFC portions from the test and the new functionality. Ready for review.

joker-eph commented on mlir/include/mlir/Tools/mlir-opt/MlirOptMain.h:
```diff
@@ -173,6 +173,10 @@ class MlirOptMainConfig {
   bool shouldVerifyRoundtrip() const { return verifyRoundtripFlag; }
 
+  /// Reproducer file generation (no crash required).
+  StringRef getReproduerFilename() const { return reproduerFilenameFlag; }
```
```suggestion
  StringRef getReproducerFilename() const { return reproducerFilenameFlag; }
```

plotfi replied:
Oops. My auto-complete must have consistently gotten this one wrong. Good catch. :-)

joker-eph commented on mlir/include/mlir/Tools/mlir-opt/MlirOptMain.h:
```diff
@@ -228,6 +232,12 @@ class MlirOptMainConfig {
   bool verifyRoundtripFlag = false;
+
+  /// The reproducer output filename (no crash required).
+  std::string reproduerFilenameFlag = "";
```
```suggestion
  std::string reproducerFilenameFlag = "";
```

joker-eph commented on mlir/include/mlir/Tools/mlir-opt/MlirOptMain.h:
```diff
@@ -228,6 +232,12 @@ class MlirOptMainConfig {
+  /// The reproducer output filename (no crash required).
+  std::string reproduerFilenameFlag = "";
+
+  /// Generate a reproducer file at reproduerFilename (no crash required).
```
```suggestion
  /// Generate a reproducer file at reproducerFilename (no crash required).
```

joker-eph commented on mlir/lib/Pass/Pass.cpp:
```diff
@@ -382,16 +382,23 @@ StringRef OpPassManager::getOpAnchorName() const {
 }
+void OpPassManager::printAsTextualPipeline(raw_ostream &os) const {
+  StringRef anchorName = getOpAnchorName();
+  auto *begin = MutableArrayRef<std::unique_ptr<Pass>>{impl->passes}.begin();
+  auto *end = MutableArrayRef<std::unique_ptr<Pass>>{impl->passes}.end();
+  const llvm::iterator_range<OpPassManager::pass_iterator> passes = {begin,
+                                                                     end};
```
Wouldn't this build?

```suggestion
  const llvm::iterator_range<OpPassManager::pass_iterator> passes(impl->passes}.begin(), impl->passes}.end());
```

plotfi replied:
This one was pretty problematic to get building, keep getting build errors like:

`....llvm/include/llvm/ADT/iterator.h:328:70: error: no matching function for call to ‘llvm::iterator_adaptor_base<llvm::pointee_iterator<std::unique_ptr<mlir::Pass>*>, std::unique_ptr<mlir::Pass>*, std::random_access_iterator_tag, mlir::Pass, long int, mlir::Pass*, mlir::Pass&>::iterator_adaptor_base(__gnu_cxx::__normal_iterator<std::unique_ptr<mlir::Pass>*, std::vector<std::unique_ptr<mlir::Pass>, std::allocator<std::unique_ptr<mlir::Pass> > > >)’
  328 |       : pointee_iterator::iterator_adaptor_base(std::forward<U &&>(u)) {}`

plotfi replied:
This works however:

```
::printAsTextualPipeline(os, anchorName, {
   MutableArrayRef<std::unique_ptr<Pass>>{impl->passes}.begin(),
   MutableArrayRef<std::unique_ptr<Pass>>{impl->passes}.end()
});
```

joker-eph commented on mlir/lib/Pass/Pass.cpp:
```diff
@@ -382,16 +382,23 @@ StringRef OpPassManager::getOpAnchorName() const {
+  auto *end = MutableArrayRef<std::unique_ptr<Pass>>{impl->passes}.end();
+  const llvm::iterator_range<OpPassManager::pass_iterator> passes = {begin,
+                                                                     end};
+  ::printAsTextualPipeline(os, anchorName, passes);
```
Or even directly:

```suggestion
  ::printAsTextualPipeline(os, anchorName, {impl->passes.begin(), impl->passes.end()});
```

joker-eph commented on mlir/lib/Pass/PassCrashRecovery.cpp:
```diff
@@ -38,7 +38,7 @@ namespace detail {
 struct RecoveryReproducerContext {
   RecoveryReproducerContext(std::string passPipelineStr, Operation *op,
-                            PassManager::ReproducerStreamFactory &streamFactory,
+                            mlir::ReproducerStreamFactory &streamFactory,
```
Do we really need the `mlir::` prefix? (here and below)

joker-eph commented on mlir/lib/Pass/PassCrashRecovery.cpp:
```diff
@@ -283,19 +289,7 @@ void PassCrashReproducerGenerator::finalize(Operation *rootOp,
-  if (!impl->activeContexts.empty())
-    impl->activeContexts.back()->disable();
-
+static std::string produceTextualPipelineWithScope(Pass *pass, Operation *&op) {
```
Why is the `op` pointer passed by reference?

plotfi replied:
Because the while-loop in the function updates it. The refactor was breaking before I passed by reference until I figured this out. I will see if I can drop refactoring out the produceTextualPipelineWithScope function however.

joker-eph replied:
OK I see. If you have to keep this, please add a comment before the function describing the update.

joker-eph commented on mlir/lib/Tools/mlir-opt/MlirOptMain.cpp:
```diff
@@ -151,6 +151,21 @@ struct MlirOptMainConfigCLOptions : public MlirOptMainConfig {
+        llvm::cl::desc("Generate a reproducer at"
+                       " --mlir-reproducer-filename=<filename> "
+                       " (no crash required)"),
+        cl::location(generateReproducerFlag), cl::init(false));
```
Can we just use one option?

`--mlir-generate-reproducer=<file>` 

If it is set we generate the repro using the provided path.
(`cl::opt` exposes `getNumOccurences()`)

plotfi replied:
That is ok with me. Do you think it matters if we generate the same file path for a crash reproducer versus not? I wasn't completely certain about this, so thats why I did a separate cl::opt flag file.

joker-eph replied:
I don't follow what you're asking about here?

You're introducing 2 new options, I'm saying you can introduce one only. But that won't conflict with the `mlir-pass-pipeline-crash-reproducer` option right?

plotfi replied:
Oh I see, I understand now.

joker-eph commented on mlir/lib/Tools/mlir-opt/MlirOptMain.cpp:
```diff
@@ -384,6 +399,13 @@ performActions(raw_ostream &os,
+  if (config.shouldGenerateReproducer()) {
+    StringRef anchorName = pm.getAnyOpAnchorName();
+    const auto &passes = pm.getPasses();
+    makeReproducer(anchorName, passes, op.get(), config.getReproduerFilename());
```
```suggestion
    makeReproducer(anchorName, passes, op.get(), config.getReproducerFilename());
```

plotfi commented:
@joker-eph Updated based on your feedback.
